### ISSUE TITLE
Periodic loss fluctuation for multi-GPUs

### ISSUE THREAD

**Issue by dorucia** [NVIDIA/DALI#1966]
Hi,

I'm using PyTorch + DALI + tfrecord with 8 GPUs.
I don't use APEX.

When I train/test my model, I get some periodic loss fluctuation like this.
![image](https://user-images.githubusercontent.com/30884952/81974584-a01ebe80-95da-11ea-9932-d17b5129df4b.png)

Its fluctuation period is 8 epochs which is the same as the number of GPUs.

I'm not sure why it happens.
I describe how I process and load the data below.

To prepare tfrecord files, I do like this.

1. Shuffle whole dataset
2. Save N=10 data samples per each tfrecord file
3. The number of tfrecord file is about 300

When I load tfrecord files, I do like this.

1. Load tfrecord and index files.
2. Shuffle the ordering
```
tmp_list = list(zip(tfrecord, tfrecord_idx))
random.shuffle(tmp_list)
tfrecord, tfrecord_idx = zip(*tmp_list)
tfrecord = list(tfrecord)
tfrecord_idx = list(tfrecord_idx)
```

Then I define the TFRecordPipeline like this.
![image](https://user-images.githubusercontent.com/30884952/81974795-ea07a480-95da-11ea-92de-4906ea8b777e.png)
where
batch_size = 64 (per GPU)
num_workers = 8
num_gpus = 8
shuffle = True for training and False for testing
mode = Train or Test
random_crop = False for now.

And the TFRecordPipeline class looks like this.

```
class TFRecordPipeline(Pipeline):
    def __init__(self, tfrecord, tfrecord_idx, batch_size, num_threads, device_id, shard_id, num_shards, shuffle, mode, random_crop):
        super(TFRecordPipeline, self).__init__(batch_size, num_threads, device_id, seed=12+device_id, exec_async=False, exec_pipelined=False)

        self.input = ops.TFRecordReader(path = tfrecord, index_path = tfrecord_idx, features = {defined based on our data},
                        shard_id = shard_id, num_shards = num_shards, random_shuffle = shuffle,
                        initial_fill = 4096, read_ahead = True)
        ...
```
I set `exec_async=False, exec_pipelined=False` since I use custom python operator.

For iterator, it is defined like this.

```
train_iter = DALIGenericIterator(data_loader, ['our data'], int(num_train_data / num_gpus))
test_iter = DALIGenericIterator(data_loader, ['our data'], int(num_test_data / num_gpus), fill_last_batch=False, last_batch_padded=True)
```

For accumulating training/testing loss, I use this function.

![image](https://user-images.githubusercontent.com/30884952/81975973-b0379d80-95dc-11ea-9e54-5b5167cc53ac.png)

I want to know why I get the fluctuation.
Thank you.

**Comment by JanuszL** [comment]
Hi,
I suspect it could be an effect that when the number of shards doesn't divide a dataset size padding gets confused. Can you try to turn on - `stick_to_shard` option and check if that changes anything?
This limitation will be addressed in https://github.com/NVIDIA/DALI/pull/1831.

**Comment by dorucia** [comment]
I turned on `stick_to_shard` but it did not help.

**Comment by JanuszL** [comment]
Does it happen without DALI as well?

**Comment by dorucia** [comment]
I don't have other way to load tfrecord from PyTorch without DALI.
So I cannot confirm this when I use tfrecord.

Without using tfrecord, when I directly loaded data from file, I did not see this issue.

**Comment by JanuszL** [comment]
> Without using tfrecord, when I directly loaded data from file, I did not see this issue.

But still using DALI? I'm trying to narrow down if this is the problem of the data container of DALI itself.

**Comment by dorucia** [comment]
No. I don't use DALI to load the data from file directly.

Maybe I can try ExternalSource in DALI but it think will take some time to fix the code.

**Comment by dorucia** [comment]
So in short, I encounter this issue when I started to use tfrecord+DALI.

**Comment by JanuszL** [comment]
@dorucia - I haven't seen such behavior until now, but it doesn't necessarily mean that there is no problem in DALI.
If you could provide a self-contained repro that we can run our side to reproduce this behavior it would be great. If not I can only ask you to do more experiments to help us narrow down the problem.

**Comment by dorucia** [comment]
I see.
I will dig from my side first.

**Comment by dorucia** [comment]
Oh, I figured out that I made a mistake when turning on the `stick_to_shard`.

Now, with the `stick_to_shard` option on, the periodic fluctuation is gone.

**Comment by JanuszL** [comment]
So keep an eye on https://github.com/NVIDIA/DALI/pull/1831, and test nightly build when it is merged.

**Comment by dorucia** [comment]
Great, thank you.
Will do that.

**Comment by dorucia** [comment]
Just in case,

I read a PyTorch document for multiprocessing. https://pytorch.org/docs/master/data.html#randomness-in-multi-process-data-loading

It says other libraries such as numpy's random number is copied to other workers.
So to deal with this issue, one has to use worker_init_fn to set the numpy random seed.

It might or might not be related to this issue depending on how random seed is set in DALI.

**Comment by JanuszL** [comment]
> So to deal with this issue, one has to use worker_init_fn to set the numpy random seed.

It is not the case for DALI. You can manually set seed or relay on the one generated automatically which should differ between processes. Even if you use the same seed in all processes it should not make any difference as each DALI instance will work on different data shards so data should not overlap.

**Comment by dorucia** [comment]
It seems the issue is resolved with the recently merged #1831.
Thanks.

Before it is merged, I use `stick_to_shard=True` as you suggested.
Now, I use `stick_to_shard=False`.
What kind of difference do we expect by changing this to False?
Will it impact the accuracy of the trained network?

**Comment by JanuszL** [comment]
> What kind of difference do we expect by changing this to False?

It tells each DALI pipeline to go through all data shards or stick to one through the training. It will affect normalization, so setting this to False should yield a bit better accuracy.

**Comment by dorucia** [comment]
Okay, thanks.
We can close the issue.

**Comment by dorucia** [comment]
Sorry. I made a mistake.
This problem is still here.

**Comment by JanuszL** [comment]
To use the new implementation from https://github.com/NVIDIA/DALI/pull/1831 you need to pass `reader_name` and set `fill_last_batch=False` the iterator, and set `pad_last_batch=True` in the reader.

**Comment by dorucia** [comment]
Hi, I tried what you suggested, however, it still has the same problem.

I think I found a root cause of this problem.

I read the sharding document of DALI (master branch).
From the equation
`floor(((id + epoch_num) % num_shards + 1) * dataset_size / num_shards) - floor(((id + epoch_num) % num_shards) * dataset_size / num_shards)`,
I can see that each DALI pipeline will get the same shard periodically.
At epoch=0 and epoch=num_shards, the allocate shard is the same.
Therefore, every num_shards epoch, the data distribution in each GPU is the same.

I think it is better to do something like this.
At every epoch, we randomly shuffle the shard id which will be corresponded to the DALI pipeline.
In addition, currently shard_id parameter only takes int.
For an ideal case, I think it should take a list of int.

Let me give you an example.

Example)
There are 2 GPUs and we want to divide a dataset into 6 shards.

At epoch=0
GPU#1: shard [1,2,3]
GPU#2: shard [4,5,6]

At epoch=1: randomly shuffle shard id
GPU#1: shard [5,3,2]
GPU#2: shard [1,6,4]

...

In this way, we can achieve global shuffling.
For local shuffling, we can use buffer based shuffling which is already provided in the DALI framework.

For more discussion about sharding, I refer https://datascience.stackexchange.com/questions/16318/what-is-the-benefit-of-splitting-tfrecord-file-into-shards

**Comment by dorucia** [comment]
I found a similar request.
I pasted it here for your reference.

<img width="737" alt="image" src="https://user-images.githubusercontent.com/30884952/84252853-01e02480-aac4-11ea-9272-5d5a07dbce1d.png">

The second paragraph is the same as what I thought in the above 2 GPU example.

**Comment by JanuszL** [comment]
Closing as stalled.
